diff --git a/src/i3.mk b/src/i3.mk
index 36a24c8..395b4cf 100644
--- a/src/i3.mk
+++ b/src/i3.mk
@@ -54,12 +54,12 @@ src/config_parser.o: src/config_parser.c $(i3_HEADERS_DEP) i3-config-parser.stam
 
 i3-command-parser.stamp: generate-command-parser.pl parser-specs/commands.spec
 	echo "[i3] Generating command parser"
-	(cd include; ../generate-command-parser.pl --input=../parser-specs/commands.spec --prefix=command)
+	(cd $(TOPDIR)/include; ../generate-command-parser.pl --input=../parser-specs/commands.spec --prefix=command)
 	touch $@
 
 i3-config-parser.stamp: generate-command-parser.pl parser-specs/config.spec
 	echo "[i3] Generating config parser"
-	(cd include; ../generate-command-parser.pl --input=../parser-specs/config.spec --prefix=config)
+	(cd $(TOPDIR)/include; ../generate-command-parser.pl --input=../parser-specs/config.spec --prefix=config)
 	touch $@
 
 i3: libi3.a $(i3_OBJECTS)
